dcpandas/df_builder.py

```diff
@@ -0,0 +1,292 @@
+      place (`str`): The dcid of Place to query for.
+      stat_var (`str`): The dcid of the StatisticalVariable.
+    Returns:
+      A pandas Series with Place IDs as the index, and Observed statistics as values.
+    """
+    return pd.Series(dc.get_stat_series(place, stat_var))
+
+
+def _group_stat_all_by_obs_options(places, stat_vars, mode):
+    """Groups the result of `get_stat_all` by Observation options for time series.
+    
+    Args:
+      places (`str` or `iterable` of `str`): The dcids of Places to query for.
+      stat_vars (`Iterable` of `str`): The dcids of the StatisticalVariables.
+      mode (`str`): "series" to output time series grouped by Observation options, or
+        "covariates" to output latest Observations.
+    Returns:
+      A pandas Series with Place IDs as the index, and Observed statistics as values.
+
+    Raises:
+      ValueError: If the payload returned by the Data Commons REST API is
+        malformed.
+    """
+    kseries = "series"
+    kcov = "covariates"
+
+    if mode == kseries:
+        if len(stat_vars) != 1:
+            raise ValueError(
+                'When `mode=series`, only one StatisticalVariable for `stat_vars` is allowed.'
+            )
+        res = collections.defaultdict(list)
+    elif mode == kcov:
+        res = collections.defaultdict(lambda: collections.defaultdict(list))
+    else:
+        raise ValueError(
+            'Value of `mode` must be one of ("series", "covariates")')
+
+    stat_all = dc.get_stat_all(places, stat_vars)
+    for place, place_data in stat_all.items():
+        if not place_data:
+            continue
+        for stat_var, stat_var_data in place_data.items():
+            if not stat_var_data:
+                continue
+            for source_series in stat_var_data['sourceSeries']:
+                time_series = source_series['val']
+                # Create a hashable for Observation options.
+                obs_options = (('measurementMethod',
+                                source_series.get('measurementMethod')),
+                               ('observationPeriod',
+                                source_series.get('observationPeriod')),
+                               ('unit', source_series.get('unit')),
+                               ('scalingFactor',
+                                source_series.get('scalingFactor')))
+                if mode == kseries:
+                    res[obs_options].append(
+                        dict({'place': place}, **time_series))
+                elif mode == kcov:
+                    date = max(time_series)
+                    res[stat_var][obs_options].append({
+                        'place': place,
+                        'date': date,
+                        'val': time_series[date]
+                    })
+    if mode == kseries:
+        return dict(res)
+    elif mode == kcov:
+        return {k: dict(v) for k, v in res.items()}
+
+
+def _time_series_pd_input(places, stat_var):
```

Comment:
since this is a private helper, consider renaming this to be more of a verb. what is it doing to the inputs? why would you call this function? it's also non-obvious that the ranking/selection happens here. it should be very obvious what the transformation is from the call site.